File: clients/php/Debugging/ColoredVertex.php
<?php

namespace Debugging {
    require_once 'Debugging/Color.php';
    require_once 'Model/Vec2.php';
    require_once 'Stream.php';

    /**
     * Point + color
     */
    class ColoredVertex
    {
        /**
         * Position
         */
        public \Model\Vec2 $position;
        /**
         * Color
         */
        public \Debugging\Color $color;
    
        function __construct(\Model\Vec2 $position, \Debugging\Color $color)
        {
            $this->position = $position;
            $this->color = $color;
        }
    
        /**
         * Read ColoredVertex from input stream
         */
        public static function readFrom(\InputStream $stream): ColoredVertex
        {
            $position = \Model\Vec2::readFrom($stream);
            $color = \Debugging\Color::readFrom($stream);
            return new ColoredVertex($position, $color);
        }
        
        /**
         * Write ColoredVertex to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $this->position->writeTo($stream);
            $this->color->writeTo($stream);
        }
    }
}

File: clients/cpp/debugging/ColoredVertex.cpp
#include "ColoredVertex.hpp"

namespace debugging {

ColoredVertex::ColoredVertex(model::Vec2 position, debugging::Color color) : position(position), color(color) { }

// Read ColoredVertex from input stream
ColoredVertex ColoredVertex::readFrom(InputStream& stream) {
    model::Vec2 position = model::Vec2::readFrom(stream);
    debugging::Color color = debugging::Color::readFrom(stream);
    return ColoredVertex(position, color);
}

// Write ColoredVertex to output stream
void ColoredVertex::writeTo(OutputStream& stream) const {
    position.writeTo(stream);
    color.writeTo(stream);
}

// Get string representation of ColoredVertex
std::string ColoredVertex::toString() const {
    std::stringstream ss;
    ss << "ColoredVertex { ";
    ss << "position: ";
    ss << position.toString();
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

}